docs/src/internals.md

```diff
@@ -0,0 +1,256 @@
+and wrapping each symbol in a `Term` constructor:
+
+```julia-repl
+julia> @macroexpand @formula(y ~ 1 + a*b)
+:(Term(:y) ~ ConstantTerm(1) + Term(:a) + Term(:b) + Term(:a) & Term(:b))
+```
+
+Calling this stage "macro time" is a bit of a misnormer because much of the
+action happens when the expression returned by the `@formula` macro is
+evaluated.  At this point, the `Term`s are combined to create higher-order terms
+via overloaded methods for `~`, `+`, and `&`:
+
+```@repl 1
+using StatsModels; # hide
+dump(Term(:a) & Term(:b))
+dump(Term(:a) + Term(:b))
+dump(Term(:y) ~ Term(:a))
+```
+
+!!! note
+    
+    As always, you can introspect which method is called with
+
+    ```julia
+    julia> @which Term(:a) & Term(:b)
+    &(terms::AbstractTerm...) in StatsModels at /home/dave/.julia/dev/StatsModels/src/terms.jl:224
+    ```
+
+The reason that the actual construction of higher-order terms is done after the
+macro is expanded is that it makes it much easier to create a formula
+programatically:
+
+```@repl 1
+f = Term(:y) ~ sum(Term(s) for s in [:a, :b, :c])
+f == @formula(y ~ a + b + c)
+```
+
+The major exception to this is that non-DSL calls **must** be specified using
+the `@formula` macro.  The reason for this is that non-DSL calls are "captured"
+and turned into anonymous functions that can be evaluated elementwise, which has
+to happen at compile time.  For instance, the call to `log` in `@formula(y ~
+log(a+b))` is converted into the anonymous function `(a,b) -> log(a+b)`.
+
+### Schema time
+
+The next phase of life for a formula begins when a _schema_ for the data becomes
+available.  A schema is a mapping from data columns to a concrete term
+type---either a `ContinuousTerm` or a `CategoricalTerm`---which represents all
+the summary information about a data column necessary to create a model matrix
+from that column.
+
+A schema is computed with the `schema` function.  By default, it will create a
+schema for every column in the data:
+
+```@repl 1
+using DataFrames    # for pretty printing---any Table will do
+df = DataFrame(y = rand(9), a = [1:9;], b = rand(9), c = repeat(["a","b","c"], 3))
+schema(df)
+```
+
+However, if a term (including a `FormulaTerm`) is provided, the schema will be
+computed based only on the necessary variables:
+
+```@repl 1
+schema(@formula(y ~ 1 + a), df)
+schema(Term(:a) + Term(:b), df)
+```
+
+Once a schema is computed, it's _applied_ to the formula with
+[`apply_schema`](@ref).  Among other things, this _instantiates_ placeholder
+terms: 
+* `Term`s become `ContinuousTerm`s or `CategoricalTerm`s
+* `ConstantTerm`s become `InterceptTerm`s
+* Tuples of terms [`MatrixTerm`](@ref)s where appropriate to explicitly indicate
+  they should be concatenated into a single model matrix
+
+```@repl 1
+f = @formula(y ~ 1 + a + b + c)
+typeof(f)
+f = apply_schema(f, schema(f, df))
+typeof(f)
+```
+
+This transformation is done by calling `apply_schema(term, schema, ModelType)`
```

Comment:
Should be `modeltype` or `MT` to make it clear that's not an actual type, but an argument to provide (same below).